clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java

```diff
@@ -304,10 +306,23 @@ public void poll(long now, long remainingMs) {
             // When group management is used, metadata wait is already performed for this scenario as
             // coordinator is unknown, hence this check is not required.
             if (metadata.updateRequested() && !client.hasReadyNodes()) {
-                boolean metadataUpdated = client.awaitMetadataUpdate(remainingMs);
-                if (!metadataUpdated && !client.hasReadyNodes())
-                    return;
+                long iterations = (remainingMs - remainingMs % heartbeat.interval()) / heartbeat.interval();
```

Comment:
From my understanding, whenever retrieveOffsetsByTime is called, awaitMetadataUpdate will block. So I thought if that was the case, I will segment the poll times into several intervals to ensure that heartbeat stays on time.

Reply:
The threading model may not be entirely clear. The consumer has two threads: the foreground user thread and the heartbeat thread (`AbstractCoordinator.HeartbeatThread`). Since the logic inside `Fetcher` and `ConsumerCoordinator` are both executing in the same thread, there is no similar problem. And since heartbeats aren't executing in this path at all, there is no contention with the heartbeat thread either. So there shouldn't be any need to change anything here.